lib/puppet/pops/types/type_mismatch_describer.rb

```diff
@@ -598,7 +620,16 @@ def describe_struct_signature(params_struct, param_hash, missing_ok = false)
         value = param_hash[name]
         value_type = elem.value_type
         if param_hash.include?(name)
-          result << describe(value_type, TypeCalculator.singleton.infer_set(value), [ParameterPathElement.new(name)]) unless value_type.instance?(value)
+          if Puppet::Pops::Types::TypeFactory.deferred.implementation_class == value.class
+            if (df_return_type = get_deferred_function_return_type(value))
+              result << describe(value_type, df_return_type, [ParameterPathElement.new(name)]) unless matching_return_type?(df_return_type, value_type)
+            else
+              Puppet.warning _("Deferred function %{function_name} has no return_type, unable to guarantee value type during compilation.") %
```

Comment:
Maybe `Puppet.warn_once` to avoid spamming the log

Reply:
`warn_once` requires that there be a `kind` provided, which must match `deprecations, undefined_variables, undefined_resources`. This doesn't really match any of those kinds. Do you think it makes sense to add a kind for this particular usage? I could see some sort of general idea that we want to deprecate functions that don't define `return_types`...

Reply:
I think the `kind` must match the list if you want to be able to disable them via `Puppet[:disable_warnings] = 'deprecations`. But you can use kinds that aren't in the list like https://github.com/puppetlabs/puppet/blob/c0f692088789217a35486714de2e9d53bfaf6622/lib/puppet/gettext/module_translations.rb#L20 I could go either way about deprecation vs something else. Certainly seems like a best practice to specify the return type.